File: python/tests/guardrails_nudge_test.py
"""Tests for nudge system."""

from pedro_agentware.middleware.guardrails.nudge import (
    NudgeKind,
    prerequisite_nudge,
    retry_nudge,
    step_nudge,
    unknown_tool_nudge,
)


class TestNudgeKind:
    def test_retry_kind(self):
        assert NudgeKind.RETRY.value == "retry"

    def test_unknown_tool_kind(self):
        assert NudgeKind.UNKNOWN_TOOL.value == "unknown_tool"

    def test_step_kind(self):
        assert NudgeKind.STEP.value == "step"

    def test_prerequisite_kind(self):
        assert NudgeKind.PREREQUISITE.value == "prerequisite"


class TestRetryNudge:
    def test_basic(self):
        tools = ["get_weather", "echo", "search"]
        nudge = retry_nudge("some text", tools)

        assert nudge.role == "user"
        assert nudge.kind == NudgeKind.RETRY
        assert nudge.tier == 0
        assert "get_weather" in nudge.content

    def test_empty_tools(self):
        nudge = retry_nudge("text", [])
        assert "(no tools available)" in nudge.content


class TestUnknownToolNudge:
    def test_basic(self):
        tools = ["echo", "search"]
        nudge = unknown_tool_nudge("nonexistent", tools)

        assert nudge.role == "user"
        assert nudge.kind == NudgeKind.UNKNOWN_TOOL
        assert "nonexistent" in nudge.content


class TestStepNudge:
    def test_tier_1(self):
        pending = ["validate", "prepare"]
        nudge = step_nudge("submit", pending, 1)

        assert nudge.kind == NudgeKind.STEP
        assert nudge.tier == 1

    def test_tier_2(self):
        pending = ["validate", "prepare"]
        nudge = step_nudge("submit", pending, 2)

        assert nudge.tier == 2

    def test_tier_3(self):
        pending = ["validate"]
        nudge = step_nudge("submit", pending, 3)

        assert nudge.tier == 3

    def test_tier_clamping_below_min(self):
        nudge = step_nudge("submit", ["validate"], 0)
        assert nudge.tier == 1

    def test_tier_clamping_above_max(self):
        nudge = step_nudge("submit", ["validate"], 10)
        assert nudge.tier == 3


class TestPrerequisiteNudge:
    def test_basic(self):
        missing = ["authenticate", "validate"]
        nudge = prerequisite_nudge("submit", missing)

        assert nudge.kind == NudgeKind.PREREQUISITE
        assert nudge.tier == 0
        assert "authenticate" in nudge.content

File: typescript/tests/guardrails-nudge.test.ts
import {
  NudgeKind,
  retryNudge,
  unknownToolNudge,
  stepNudge,
  prerequisiteNudge,
} from "../src/middleware/guardrails/nudge";

describe("NudgeKind", () => {
  it("should have correct values", () => {
    expect(NudgeKind.RETRY).toBe("retry");
    expect(NudgeKind.UNKNOWN_TOOL).toBe("unknown_tool");
    expect(NudgeKind.STEP).toBe("step");
    expect(NudgeKind.PREREQUISITE).toBe("prerequisite");
  });
});

describe("retryNudge", () => {
  it("should create a retry nudge", () => {
    const tools = ["get_weather", "echo", "search"];
    const nudge = retryNudge("some text", tools);

    expect(nudge.role).toBe("user");
    expect(nudge.kind).toBe(NudgeKind.RETRY);
    expect(nudge.tier).toBe(0);
    expect(nudge.content).toContain("get_weather");
  });

  it("should handle empty tools list", () => {
    const nudge = retryNudge("text", []);
    expect(nudge.content).toContain("(no tools available)");
  });
});

describe("unknownToolNudge", () => {
  it("should create an unknown tool nudge", () => {
    const tools = ["echo", "search"];
    const nudge = unknownToolNudge("nonexistent", tools);

    expect(nudge.role).toBe("user");
    expect(nudge.kind).toBe(NudgeKind.UNKNOWN_TOOL);
    expect(nudge.content).toContain("nonexistent");
  });
});

describe("stepNudge", () => {
  it("should create tier 1 nudge", () => {
    const pending = ["validate", "prepare"];
    const nudge = stepNudge("submit", pending, 1);

    expect(nudge.kind).toBe(NudgeKind.STEP);
    expect(nudge.tier).toBe(1);
  });

  it("should create tier 2 nudge", () => {
    const pending = ["validate", "prepare"];
    const nudge = stepNudge("submit", pending, 2);

    expect(nudge.tier).toBe(2);
  });

  it("should create tier 3 nudge", () => {
    const pending = ["validate"];
    const nudge = stepNudge("submit", pending, 3);

    expect(nudge.tier).toBe(3);
  });

  it("should clamp tier below minimum", () => {
    const nudge = stepNudge("submit", ["validate"], 0);
    expect(nudge.tier).toBe(1);
  });

  it("should clamp tier above maximum", () => {
    const nudge = stepNudge("submit", ["validate"], 10);
    expect(nudge.tier).toBe(3);
  });
});

describe("prerequisiteNudge", () => {
  it("should create a prerequisite nudge", () => {
    const missing = ["authenticate", "validate"];
    const nudge = prerequisiteNudge("submit", missing);

    expect(nudge.kind).toBe(NudgeKind.PREREQUISITE);
    expect(nudge.tier).toBe(0);
    expect(nudge.content).toContain("authenticate");
  });
});